gui/src/app/pages/HomePage/HomePage.tsx

```diff
@@ -124,30 +124,47 @@ const LeftView: FunctionComponent<LeftViewProps> = () => {
   );
 };
 
-const DataEditor: FunctionComponent<{}> = () => {
+const DataEditor: FunctionComponent<unknown> = () => {
   const { data, update } = useContext(ProjectContext);
+  const defaultSourceDesc = "";
 
   const dataIsEdited = useMemo(() => {
     return data.dataFileContent !== data.ephemera.dataFileContent;
   }, [data.dataFileContent, data.ephemera.dataFileContent]);
 
+  const sourceDesc = useMemo(() => {
+    if (dataIsEdited) return defaultSourceDesc;
+    switch (data.meta.dataSource) {
+      case DataSource.GENERATED_BY_PYTHON: {
+        return "data.py";
+      }
+      case DataSource.GENERATED_BY_R: {
+        return "data.R";
+      }
+      case DataSource.GENERATED_BY_STALE_PYTHON: {
+        return "a prior version of data.py.";
+      }
+      case DataSource.GENERATED_BY_STALE_R: {
+        return "a prior version of data.R.";
+      }
+      default:
+        return defaultSourceDesc;
```

Comment:
Could this be one of our "unreachable" calls?

Reply:
It is reachable though--this branch handles the case where `dataSource` is `undefined`.

I briefly considered wrapping this up further to remove the branch in the `dataMessage` memoization right below, but it felt like that was maybe a bit too cute.

Reply:
Ah, fair point. I think it's the defaultSourceDesc string variable that bothers me -- would it make sense to make this a `string | undefined` rather than the empty string being special?

Reply:
I suppose making it possibly `undefined` means we always have to keep the explicit branch check in constructing the message, vs. we could set it up so that the `sourceDesc` always has the right string and then there's no need to branch in l156-157 below.

But I have very little emotional investment in this choice :)

Reply:
I was assuming that an empty toolbar would not be visible to the user.

Apart from that, this seems like a pretty minor distinction and I'm happy to do it whichever way you prefer.